server/src/main/java/org/elasticsearch/common/logging/LoggingOutputStream.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+/**
+ * A stream whose output is sent to the configured logger, line by line.
+ */
+class LoggingOutputStream extends OutputStream {
+    /** The starting length of the buffer */
+    static final int DEFAULT_BUFFER_LENGTH = 1024;
+
+    class Buffer {
+
+        /** The buffer of bytes sent to the stream */
+        byte[] bytes = new byte[DEFAULT_BUFFER_LENGTH];
+
+        /** Number of used bytes in the buffer */
+        int used = 0;
+    }
+
+    // each thread gets its own buffer so messages don't get garbled
+    ThreadLocal<Buffer> threadLocal = ThreadLocal.withInitial(Buffer::new);
+
+    private final Logger logger;
+
+    private final Level level;
+
+    LoggingOutputStream(Logger logger, Level level) {
+        this.logger = logger;
+        this.level = level;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        if (threadLocal == null) {
+            throw new IOException("buffer closed");
+        }
+        if (b == 0) return;
+        if (b == '\n') {
```

Comment:
so this looks fine in plaintext logs, but might end up harder to analyze with JSON logs. Why not adding `\n` to the buffer too?

Reply:
That would cause double the lines in the log. Think of it as making this equivalent to most of our logging calls do not contain a newline in the message. Log4j inserts newlines per message emitted.

Reply:
makes sense, thank you